File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) return head;
        Node p = head;
        while (p != null) {
            Node headNext = p.next;
            p.next = new Node(p.val);
            p.next.next = headNext;
            p = headNext;
        }
        p = head;
        while (p != null) {
            if (p.random != null) {
                p.next.random = p.random.next;
            }
            p = p.next.next;
        }
        p = head;
        Node newHead = head.next;
        Node p1 = newHead;
        while (p1.next != null) {
            p.next = p.next.next;
            p = p.next;
            p1.next = p1.next.next;
            p1 = p1.next;
        }
        p.next = p.next.next;
        return newHead;
    }
}

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
class Solution {
public:
    Node* copyRandomList(Node* head) {
        if (head == nullptr) {
            return head;
        }
        
        Node* cur = head;
        
        while (cur != nullptr) {
            Node* next_node = cur->next;
            cur->next = new Node(cur->val);
            cur->next->next = next_node;
            cur = next_node;
        }
        
        cur = head;
        
        while (cur != nullptr) {
            if (cur->random != nullptr) {
                cur->next->random = cur->random->next;
            }
            cur = cur->next->next;
        }
        
        cur = head;
        Node* new_head = cur->next;
        Node* p = cur->next;
        
        while (p->next != nullptr) {
            cur->next = cur->next->next;
            cur = cur->next;
            p->next = p->next->next;
            p = p->next;
        }
        
        cur->next = cur->next->next;
        
        return new_head;
    }
};